server/routes/auth/providers/google.ts

```diff
@@ -57,37 +58,74 @@ if (env.GOOGLE_CLIENT_ID && env.GOOGLE_CLIENT_SECRET) {
         ) => void
       ) {
         try {
+          const state = req.cookies.get("state");
+          const host = state ? parseState(state).host : req.hostname;
+          const appDomain = parseDomain(host);
+
+          let result;
           const domain = profile._json.hd;
 
-          if (!domain) {
-            throw GoogleWorkspaceRequiredError();
+          // Existence of domain means this is a Google Workspaces account
+          // so we'll attempt to provision an account (team and user)
+          if (domain) {
+            const subdomain = domain.split(".")[0];
+            const teamName = capitalize(subdomain);
+
+            result = await accountProvisioner({
+              ip: req.ip,
+              team: {
+                name: teamName,
+                domain,
+                subdomain,
+              },
+              user: {
+                email: profile.email,
+                name: profile.displayName,
+                avatarUrl: profile.picture,
+              },
+              authenticationProvider: {
+                name: providerName,
+                providerId: domain,
+              },
+              authentication: {
+                providerId: profile.id,
+                accessToken,
+                refreshToken,
+                scopes,
+              },
+            });
+          } else {
+            // No domain means it's a personal Gmail account
+            // We only allow sign-in to existing invites here
+            let team;
+            if (appDomain.custom) {
```

Comment:
If self-hosted we can just use the first team here to add support for that scenario